codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java

```diff
@@ -308,23 +315,93 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
         assertFalse(channel.finish());
     }
 
+    static class WriteFailingByteBuf extends UnpooledHeapByteBuf {
+        private final Error error = new Error();
+        private int untilFailure;
+
+        WriteFailingByteBuf(int untilFailure, int capacity) {
+            super(UnpooledByteBufAllocator.DEFAULT, capacity, capacity);
+            this.untilFailure = untilFailure;
+        }
+
+        @Override
+        public ByteBuf writeBytes(ByteBuf src) {
+            if (--untilFailure <= 0) {
+                throw error;
+            }
+            return super.writeBytes(src);
+        }
+
+        Error writeError() {
+            return error;
+        }
+    }
+
     @Test
     public void releaseWhenMergeCumulateThrows() {
-        final Error error = new Error();
+        WriteFailingByteBuf oldCumulation = new WriteFailingByteBuf(1, 64);
+        ByteBuf in = Unpooled.buffer().writeZero(12);
 
-        ByteBuf cumulation = new UnpooledHeapByteBuf(UnpooledByteBufAllocator.DEFAULT, 0, 64) {
+        Throwable thrown = null;
+        try {
+            ByteToMessageDecoder.MERGE_CUMULATOR.cumulate(UnpooledByteBufAllocator.DEFAULT, oldCumulation, in);
+        } catch (Throwable t) {
+            thrown = t;
+        }
+
+        assertSame(oldCumulation.writeError(), thrown);
+        assertEquals(0, in.refCnt());
+        assertEquals(1, oldCumulation.refCnt());
```

Comment:
sorry missed this before... You also need to call `oldCumulation .release()` here

Reply:
Yea sorry from me as well, should have noticed.